### ISSUE TITLE
Password for each data set

### ISSUE THREAD

**Issue by dskyeong** [maximilianh/cellBrowser#154]
Hi, I've installed cell browser 0.7.1.
Before releasing data, we need to keep it in personal.
Please, let me know how I can set the authority for each data set.
Thank you

**Comment by maximilianh** [comment]
There is no password right now built into it... the reason is that it's
hard to limit access from Javascript...

However, because the cell browser is so small, you could simply create a
separate cell browser per each project / user. Then, you could put them
into different subdirectories of your http server, and have a
username/password for each directory. Would that work for you?

Maybe if you explain your setup a little, I can try to find something more
customized. Do you have one browser per "customer" ? Are you a core unit?
What is your existing storage solution ? How did you handle this problem
until now?

thanks!
Max

**Comment by Gibbsdavidl** [comment]
I found a partial solution for this:  https://github.com/matteobrusa/Password-protection-for-static-pages

Essentially, a password is used to create a long hash, which is then used as a directory name... you create that directory (with the hash as name) and you pop the index.html (and other files) into that hashed directory.  works well.

**Comment by lijxug** [comment]
Waiting for a secured access to the website, too.
I followed what Gibbsdavidl recommended. The web page didn't redirect, wierd.

**Comment by maximilianh** [comment]
@lijxug can you explain what you are trying to do? In principle, if you want a protected URL to send to your users, you can simply create a URL that is very hard to guess. That should be sufficient. The solution by @Gibbsdavidl is not a real password, it just redirects, so if you send around URLs, you don't need the URL. 

Also, you can simply activate normal htaccess passwords in your webserver: https://www.digitalocean.com/community/tutorials/how-to-set-up-password-authentication-with-apache-on-ubuntu-14-04

**Comment by lijxug** [comment]
@maximilianh Yes, that's a partial solution. I've made some changes to that index.html and eventually made it work. 

Basically, I have constructed a webpage using your browser. Then I want to send the link directed to the browser only for my coworkers. So I need a password to protect this data set. 

Here is what I did:
clone the https://github.com/ulfaslak/Password-protection-for-static-pages repo because he fixed the sha1 loading problem. 
hosted the server with cdBuild
open the original index.html, insert the codes suggested by https://github.com/matteobrusa/Password-protection-for-static-pages/issues/22 after the <script> line.
generate the sha1 hash string and rename the original index.html as <sha1string>.html
cp index.html from the aforementioned repo to replace the original one. cp the sha1.js to the root of the website as well.

I'll admit that this is quite ugly. But it works for me. The webpage now requires a password and reveals none of the sha1 string in the url.

I will try your solution for apache later as well. But right now I am pretty satisfied by this redirecting solution. ;-p

**Comment by maximilianh** [comment]
Hey, is this for a submitted manuscript? Should we host the dataset for
you, once it's accepted? Or are you building a general single cell website?

**Comment by lijxug** [comment]
No, that's just a temporal web browser for my coworkers right now.
This project is not ready to publish yet, but we appreciate your offer and will definitely consider it in the future!

**Comment by maximilianh** [comment]
I think you're overdoing this. If it's internal-only, you don't need a
password at all. I'd just put it onto your webserver, into a subdirectory
with the name 38Q1Tb9RhT (see random.org). As long as your webserver does
not allow browsing all directories (they usually don't), this will be just
fine.
